lib/puppet/configurer.rb

```diff
@@ -181,25 +181,51 @@ def apply_catalog(catalog, options)
   # which accepts :tags and :ignoreschedules.
   def run(options = {})
     pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
+    # We create the report pre-populated with default settings for
+    # environment and transaction_uuid very early, this is to ensure
+    # they are sent regardless of any catalog compilation failures or
+    # exceptions.
+    options[:report] ||= Puppet::Transaction::Report.new("apply", nil, @environment, @transaction_uuid)
+    report = options[:report]
+    init_storage
+
+    Puppet::Util::Log.newdestination(report)
+
     begin
       Puppet.override(:http_pool => pool) do
-        run_internal(options)
+
+        # Skip failover logic if the server_list setting is empty
+        if Puppet.settings[:server_list].nil? || Puppet.settings[:server_list].empty?
+          do_failover = false;
+        else
+          do_failover = true
+        end
+        # When we are passed a catalog, that means we're in apply
+        # mode. We shouldn't try to do any failover in that case.
+        if options[:catalog].nil? && do_failover
+          found = find_functional_server()
+          server = found[:server]
+          if server.nil?
+            Puppet.warning "Could not select a functional puppet master"
+            server = [nil, nil]
+          end
+          Puppet.override(:server => server[0], :serverport => server[1]) do
+            Puppet.debug "Selected master: #{server[0]}:#{server[1]}"
+            report.master = "#{server[0]}:#{server[1]}"
```

Comment:
This only happens in the failover case. The non-failover path is on line 219. If we want to add the master info to that path we'd need to add a line there to set `report.master` from the old settings